Document kops setup (#339)
cloudposse/geodesic: docs/kops.md

# Kubernetes Operations (kops)

Kops is one of the easiest ways to get a production grade Kubernetes cluster up and running. The `kops` command line tool (cli) is like `kubectl` for clusters. It handles all the standard CRUD operations necessary to manage the complete life cycle of a cluster.

It is possible to run any number of [kops clusters](http://github.com/kubernetes/kops) within an account. Our "best practice" is to define one cluster per project directory in the `/conf` folder. Then define a `.envrc` ([direnv](https://direnv.net/)) configuration per directory.

> Any settings in this file will be automatically loaded when you `cd` in to the directory. Alternatively, they can be executed explicitly by running `direnv exec $directory $command`. This is useful when running commands as part of a CI/CD GitOps-style pipeline.

Reviewer: ```suggestion
Any settings in this file will be automatically loaded when you `cd` into the directory. Alternatively, they can be executed explicitly by running `direnv exec $directory $command`. This is useful when running commands as part of a CI/CD GitOps-style pipeline.
```

| KOPS_AVAILABILITY_ZONES                            | AWS Availability Zones (AZs) to use. Must all reside in the same region. Use an _odd_ number.  |
| KOPS_AWS_IAM_AUTHENTICATOR_ENABLED                 | Toggle IAM Authenticator support                                                               |
| KOPS_BASE_IMAGE                                    | AWS AMI base image for all EC2 instances                                                       |
| KOPS_BASTION_PUBLIC_NAME                           | Hostname that will be used for the bastion instance                                            |
| KOPS_CLOUDWATCH_DETAILED_MONITORING                | Toggle detailed CloudWatch monitoring (increases operating costs)                              |
| KOPS_CLUSTER_AUTOSCALER_ENABLED                    | Toggle the Kubernetes node autoscaler capability                                               |
| KOPS_CLUSTER_NAME                                  | Cluster base hostname (E.g. `${AWS_REGION}.${DNS_ZONE}`)                                       |

> | KOPS_DNS_ZONE                                      | Authoritative DNS Zone that will be populated automatic with hostnames                         |

Reviewer: ```suggestion
| KOPS_DNS_ZONE                                      | Authoritative DNS Zone that will be populated automatically with hostnames                         |
```

> | KOPS_FEATURE_FLAGS                                 | Enable experimental features that not available by default                                     |

Reviewer: ```suggestion
| KOPS_FEATURE_FLAGS                                 | Enable experimental features that are not available by default                                     |
```

| KOPS_KUBE_API_SERVER_AUTHORIZATION_MODE            | Ordered list of plug-ins to do authorization on secure port                                    |
| KOPS_KUBE_API_SERVER_AUTHORIZATION_RBAC_SUPER_USER | Username of the Kubernetes Super User                                                          |
| KOPS_MANIFEST                                      | The path to the manifest. Used by `build-kops-manifest`.                                       |
| KOPS_NETWORK_CIDR                                  | The network used by kubernetes for `Pods` and `Services` in the cluster                        |
| KOPS_NON_MASQUERADE_CIDR                           | A list of strings in CIDR notation that specify the non-masquerade ranges.                     |
| KOPS_PRIVATE_SUBNETS                               | Subnet CIDRs for all EC2 instances                                                             |
| KOPS_STATE_STORE                                   | S3 Bucket that will be used to store the cluster state (E.g. `s3://${AWS_REGION}.${DNS_ZONE}`) |

> | KOPS_TEMPLATE                                      | Kops manifest go-template (gomplate) that descri                                               |

Reviewer: ```suggestion
| KOPS_TEMPLATE                                      | Kops manifest go-template (gomplate) that describes the cluster                                               |
```

| KOPS_UTILITY_SUBNETS                               | Subnet CIDRs for the publically facing services (e.g. ingress ELBs)                            |

> | KUBERNETES_VERSION                                 | Version of Kubernetes to deploy. Must be compatible with the `kops` release.                   |

Reviewer: ```suggestion
| KUBERNETES_VERSION                                 | Version of Kubernetes to deploy. Must be compatible with the `kops` release.  See [kops compatibility matrix](https://github.com/kubernetes/kops#compatibility-matrix) for more details                 |
```

```bash
kops create secret sshpublickey admin -i /secrets/tf/ssh/${NAMESPACE}-${STAGE}-kops-${AWS_REGION}.pub --name $KOPS_CLUSTER_NAME
```

Run the following command to provision the AWS resources for the cluster:

```bash
kops update cluster --yes
```

All done. The `kops` cluster is now up and running.

Run the following command to validate the cluster is healthy:

```bash
kops validate cluster
```

## Operating the Cluster

Use the `kubectl` command to interact with the Kubernetes cluster. To use the `kubectl` command (_e.g._ `kubectl get nodes`, `kubectl get pods`), you need to first export the `kubecfg` configuration settings from the cluster.

Run the following command to export `kubecfg` settings needed to connect to the cluster:

```bash
kops export kubecfg
```

**IMPORTANT:** You need to run this command every time you start a new shell and before you interact with the cluster (e.g. before running `kubectl`). By default, we set the `KUBECONFIG=/dev/shm/kubecfg` (shared memory based filesystem) so that it never touches disk and is wiped out when the shell exits.

> See the documentation for [`kubecfg` settings for `kubectl`](https://github.com/kubernetes/kops/blob/master/docs/kubectl.md) for more details.

Reviewer: ```suggestion
See the documentation for [`kubecfg` settings for `kubectl`(https://github.com/kubernetes/kops/blob/master/docs/kubectl.md) for more details.
```

<details><summary>Show Example Output</summary>

Below is an example of what it should _roughly_ look like (IPs and Availability Zones may differ).

```
⨠ kops validate cluster

Validating cluster us-west-2.example.company.co

INSTANCE GROUPS
NAME			ROLE	MACHINETYPE	MIN	MAX	SUBNETS
bastions		Bastion	t2.medium	1	1	utility-${aws_region}a,utility-${aws_region}d,utility-${aws_region}c
master-${aws_region}a	Master	t2.medium	1	1	${aws_region}a
master-${aws_region}c	Master	t2.medium	1	1	${aws_region}c
master-${aws_region}d	Master	t2.medium	1	1	${aws_region}d
nodes			Node	t2.medium	2	2	${aws_region}a,${aws_region}d,${aws_region}c

NODE STATUS
NAME							                  ROLE	  READY
ip-172-20-108-58.${aws_region}.compute.internal	  node	  True
ip-172-20-125-166.${aws_region}.compute.internal  master  True
ip-172-20-62-206.${aws_region}.compute.internal	  master  True
ip-172-20-74-158.${aws_region}.compute.internal	  master  True
ip-172-20-88-143.${aws_region}.compute.internal	  node    True

Your cluster us-west-2.example.company.co is ready
```

</details>
<br>

Run the following command to list all nodes:

```bash
kubectl get nodes
```

> <details><summary>Show Exaple Output</summary>

Reviewer: ```suggestion
<details><summary>Show Example Output</summary>
```

Below is an example of what it should _roughly_ look like (IPs and Availability Zones may differ).

```
⨠ kubectl get nodes
NAME                                                STATUS   ROLES    AGE   VERSION
ip-172-20-108-58.${aws_region}.compute.internal    Ready    node     15m   v1.10.8
ip-172-20-125-166.${aws_region}.compute.internal   Ready    master   17m   v1.10.8
ip-172-20-62-206.${aws_region}.compute.internal    Ready    master   18m   v1.10.8
ip-172-20-74-158.${aws_region}.compute.internal    Ready    master   17m   v1.10.8
ip-172-20-88-143.${aws_region}.compute.internal    Ready    node     16m   v1.10.8
```

</details>
<br>

Run the following command to list all pods:

```bash
kubectl get pods --all-namespaces
```

> <details><summary>Show Exapmle Output</summary>

Reviewer: ```suggestion
<details><summary>Show Example Output</summary>
```

kube-system   kube-proxy-ip-172-20-88-143.${aws_region}.compute.internal                  1/1     Running   0          1h
kube-system   kube-scheduler-ip-172-20-125-166.${aws_region}.compute.internal             1/1     Running   0          1h
kube-system   kube-scheduler-ip-172-20-62-206.${aws_region}.compute.internal              1/1     Running   0          1h
kube-system   kube-scheduler-ip-172-20-74-158.${aws_region}.compute.internal              1/1     Running   0          1h
```

</details>
<br>
<br>

### Tips & Tricks

1. Use `kubens` to easily change your namespace context
2. Use `kubectx` to easily change between kubernetes cluster contexts
3. Use `kubeon` and `kubeoff` to enable the fancy kubernetes prompt

### Upgrade a Cluster

To upgrade the cluster or change settings (_e.g_. number of nodes, instance types, Kubernetes version, etc.):

1. Update the settings in the `.envrc` for the corresponding kops project
2. Rebuild Docker image (`make docker/build`)
3. Run `geodesic` shell (e.g. by running the wrapper script `example.company.co`)
   - assume role (`assume-role`) 

>    - change directory to the `/conf/kops` folder (or which ever project folder contains your kops configurations)

Reviewer: ```suggestion
   - change directory to the `/conf/kops` folder (or whichever project folder contains your kops configurations)
```
